src/typecheck.rs

```diff
@@ -48,20 +221,63 @@ impl<T: std::fmt::Display> std::fmt::Display for Atom<T> {
     }
 }
 
-#[derive(Default, Debug, Clone)]
-pub struct Query {
-    pub atoms: Vec<Atom<Symbol>>,
-    pub filters: Vec<Atom<Primitive>>,
+#[derive(Debug, Clone)]
+pub struct Query<T> {
+    pub atoms: Vec<Atom<T>>,
+}
+
+impl<T> Default for Query<T> {
+    fn default() -> Self {
+        Self {
+            atoms: Default::default(),
+        }
+    }
+}
+
+impl Query<Symbol> {
```

Comment:
Can we rename `typecheck.rs` to something like `query-compile.rs`?
We should also move all the typechecking related stuff like this into `typechecking.rs`

Reply:
I made an issue for this (#268). I'll do it after the PR is merged but not now to make the diff clean.